### PR TITLE
[ie/vidly] Add extractor

### PR DESCRIPTION
**Pull request by seproDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

I was unable to find a single site using vid.ly for `_WEBPAGE_TESTS`. Maybe someone else has better luck looking for those.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/vidly.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    mimetype2ext,
+    url_or_none,
+)
+from ..utils.traversal import traverse_obj
+
+
+class VidlyIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:vid\.ly/|(?:s\.)?vid\.ly/embeded.html\?(?:[^#/]+&)?link=)(?P<id>\w+)'
+    _EMBED_REGEX = [r'<script[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//vid\.ly/\w+/embed[^\'"]+)',
+                    r'<iframe[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//(?:s\.)?vid\.ly/embeded.html\?(?:[^#/]+&)?link=\w+[^\'"]+)']
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:13]: do you have any `_WEBPAGE_TESTS` we can add?

Comment (seproDev) [reply yt_dlp/extractor/vidly.py:13]: > I was unable to find a single site using vid.ly for _WEBPAGE_TESTS. Maybe someone else has better luck looking for those.

Comment (seproDev) [reply yt_dlp/extractor/vidly.py:13]: Okay found one https://www.petfinder.com/dog/gus-57378930/tn/ooltewah/furever-furkids-rescue-tn592/

File: yt_dlp/extractor/vidly.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        embed_script = self._download_webpage(
+            f'https://vid.ly/{video_id}/embed', video_id, headers={'Referer': 'https://vid.ly/'})
+        player = self._search_json(r'initCallback\(', embed_script, 'player', video_id)
+
+        if player['player'].startswith('jwplayer'):
+            return self._parse_jwplayer_data(player['config'], video_id)
+        elif player['player'] == 'vidlyplayer':
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:60]: we could do this and then unindent everything below it
```suggestion
        player_type = player.get('player') or ''
        if player_type.startswith('jwplayer'):
            return self._parse_jwplayer_data(player['config'], video_id)
        elif player_type != 'vidlyplayer':
            raise ExtractorError('Unknown player type {player_type!r}')
```

File: yt_dlp/extractor/vidly.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    mimetype2ext,
+    url_or_none,
+)
+from ..utils.traversal import traverse_obj
+
+
+class VidlyIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:vid\.ly/|(?:s\.)?vid\.ly/embeded.html\?(?:[^#/]+&)?link=)(?P<id>\w+)'
+    _EMBED_REGEX = [r'<script[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//vid\.ly/\w+/embed[^\'"]+)',
+                    r'<iframe[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//(?:s\.)?vid\.ly/embeded.html\?(?:[^#/]+&)?link=\w+[^\'"]+)']
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:13]:
```suggestion
    _VALID_URL = r'https?://(?:vid\.ly/|(?:s\.)?vid\.ly/embeded\.html\?(?:[^#]+&)?link=)(?P<id>\w+)'
    _EMBED_REGEX = [r'<script[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//vid\.ly/\w+/embed[^\'"]+)',
                    r'<iframe[^>]+\bsrc=[\'"](?P<url>(?:https?:)?//(?:s\.)?vid\.ly/embeded\.html\?(?:[^#\'"]+&)?link=\w+[^\'"]+)']
```

File: yt_dlp/extractor/vidly.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        embed_script = self._download_webpage(
+            f'https://vid.ly/{video_id}/embed', video_id, headers={'Referer': 'https://vid.ly/'})
+        player = self._search_json(r'initCallback\(', embed_script, 'player', video_id)
+
+        player_type = player.get('player') or ''
+        if player_type.startswith('jwplayer'):
+            return self._parse_jwplayer_data(player['config'], video_id)
+        elif not player_type.startswith('vidly'):
+            raise ExtractorError(f'Unknown player type {player_type!r}')
+        formats = []
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:63]: just for readability
```suggestion
            raise ExtractorError(f'Unknown player type {player_type!r}')

        formats = []
```

File: yt_dlp/extractor/vidly.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            formats.append({
+                'url': player['config']['source_hd'],
+                'format_id': 'http-hd',
+                'ext': ext,
+            })
+        # Has higher quality formats
+        formats.extend(self._extract_m3u8_formats(
+            f'https://d3fenhwk93s16g.cloudfront.net/{video_id}/hls.m3u8', video_id,
+            fatal=False, note='Trying to guess m3u8 URL') or [])
+        return {
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:81]:
```suggestion
            fatal=False, note='Requesting higher quality m3u8 formats',
            errnote='No higher quality m3u8 formats found') or [])

        return {
```

File: yt_dlp/extractor/vidly.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                'url': player['config']['source'],
+                'format_id': 'http-sd',
+                'ext': ext,
+            })
+        if traverse_obj(player, ('config', 'source_hd', {url_or_none})):
+            formats.append({
+                'url': player['config']['source_hd'],
+                'format_id': 'http-hd',
+                'ext': ext,
+            })
```
Comment (bashonly) [inline yt_dlp/extractor/vidly.py:76]:
```suggestion
        for source, fid in [('source', 'sd'), ('source_hd', 'hd')]:
            if traverse_obj(player, ('config', source, {url_or_none})):
                formats.append({
                    'url': player['config'][source],
                    'format_id': f'http-{fid}',
                    'ext': ext,
                })
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

**Pull request merged**